C/plugins/south/dht11/dht11.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * FogLAMP south service plugin
+ *
+ * Copyright (c) 2018 OSIsoft, LLC
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Amandeep Singh Arora
+ */
+#include <sstream>
+#include <iostream>
+#include <reading.h>
+#include <wiringPi.h>
+#include <dht11.h>
+#include <logger.h>
+
+#define MAX_SENSOR_READ_RETRIES	(100)
+
+using namespace std;
+
+/**
+ * Constructor for the dht11 "sensor"
+ *
+ * @param pin	The RPi pin to which DHT11 sensor's data pin is attached (e.g. pin 7 = GPIO 4, this is the default)
+ */
+DHT11::DHT11(unsigned int pin) : m_pin(pin)
+{
+	if ( wiringPiSetup() == -1 ) {
```

Comment:
This can only be called as root according to the documentation, is this true? We do not normally run as root.

Can we get any error details for the failure? Is errno set for example?

Reply:
This function tries to access /dev/mem, else /dev/gpiomem and that's normally a privileged operation. But if we allow current user to access "/dev/gpiomem" beforehand, then root access is not specifically required from plugin code.

Errno is not set and this API only returns -1 or EXIT_FAILURE. So it doesn't give more specific information.